### PR TITLE
added sdk testcases, org metadata changes ,fixed some bugs

### PR DESCRIPTION
**Pull request by keshrisohit**
added functional test cases
Setup environment with ipfs,blockchain,daemon,example service and run test cases.
Added changes related to org metadata, as group info has moved from service to organization
Now in sdk  group will contain combined info from service and org.

Right now environment is setup with fixed version of snet-daemon , will change it to build from master branch in next iteration.

### CODE REVIEW

File: packages/sdk/snet/sdk/__init__.py [hunk 1]
```diff
@@ -51,18 +59,27 @@ def __init__(
...
+        _registry_contract_address = self._config.get("registry_contract_address", None)
+        if _registry_contract_address is None:
+            self.registry_contract = get_contract_object(self.web3, "Registry.json")
+        else:
+            self.registry_contract = get_contract_object(self.web3, "Registry.json", _registry_contract_address)
 
+        self.account = Account(self.web3, config, self.mpe_contract)
 
-    def create_service_client(self, org_id, service_id, service_stub, group_name=None, payment_channel_management_strategy=PaymentChannelManagementStrategy, options=None):
+    def create_service_client(self, org_id, service_id, service_stub, group_name,
```
Comment (raamb) [inline packages/sdk/snet/sdk/__init__.py]: Is group_name mandatory now?

File: packages/sdk/snet/sdk/metadata_provider/ipfs_metadata_provider.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+from snet.snet_cli.utils_ipfs import bytesuri_to_hash, get_from_ipfs_and_checkhash
+
+import json
+
+
+class IPFSMetadataProvider(object):
+
+    def __init__(self, ipfs_client,registry_contract):
+
```
Comment (raamb) [inline packages/sdk/snet/sdk/metadata_provider/ipfs_metadata_provider.py]: Lets remove newline

File: packages/sdk/snet/sdk/metadata_provider/ipfs_metadata_provider.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.registry_contract = registry_contract
+        self._ipfs_client = ipfs_client
+
+
+    def fetch_org_metadata(self, org_id):
+        (found, id, metadata_uri, owner,members,serviceIds,repositoryIds) = self.registry_contract.functions.getOrganizationById(
+            bytes(org_id, "utf-8")).call()
+        print("dd")
```
Comment (raamb) [inline packages/sdk/snet/sdk/metadata_provider/ipfs_metadata_provider.py]: Remove print

File: packages/sdk/snet/sdk/service_client.py [hunk 4]
```diff
@@ -90,11 +99,26 @@ def _intercept_call(self, client_call_details, request_iterator, request_streami
...
 
+    def filter_existing_channels(self, new_payment_channels):
+        new_channels_to_be_added = []
+
+        # need to change this logic ,use maps to manage channels so that we can easily navigate it
+        for new_payment_channel in new_payment_channels:
+            existing_channel = False
+            for existing_payment_channel in self.payment_channels:
+                if new_payment_channel.channel_id == existing_payment_channel.channel_id:
+                    existing_channel = True
```
Comment (raamb) [inline packages/sdk/snet/sdk/service_client.py]: You can break here

File: packages/sdk/snet/sdk/service_client.py [hunk 5]
```diff
@@ -129,4 +153,6 @@ def deposit_and_open_channel(self, amount, expiration):
 
     def _get_newly_opened_channel(self, receipt):
         open_channels = self.sdk.mpe_contract.get_past_open_channels(self.sdk.account, self, receipt["blockNumber"], receipt["blockNumber"])
+        if len(open_channels) == 0:
```
Comment (raamb) [inline packages/sdk/snet/sdk/service_client.py]: Lets check whats done the snet-cli code. We may need to wait and retry a few times before we error out

File: packages/sdk/testcases/functional_tests/test_sdk_client.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+def open_first_channel(service_client):
+    channel = service_client.open_channel(123456, 33333)
+    assert channel.channel_id == 0
+    assert channel.state['nonce'] == 0
+    assert channel.state['last_signed_amount'] == 0
+
+
+def first_call_to_service_after_opening_first_channel(service_client):
+    request = examples_service_pb2.Numbers(a=20, b=3)
+    result = service_client.service.mul(request)
```
Comment (raamb) [inline packages/sdk/testcases/functional_tests/test_sdk_client.py]: Lets validate the result to be equal to 60

File: packages/sdk/testcases/functional_tests/test_sdk_client.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    channels = service_client.load_open_channels()
+
+    assert channels[0].channel_id == 0
+    assert channels[0].state['nonce'] == 0
+    assert channels[0].state['last_signed_amount'] == 1000
+
+
+def second_call_to_service_after_opening_first_channel(service_client):
+    request = examples_service_pb2.Numbers(a=20, b=3)
+    result = service_client.service.mul(request)
```
Comment (raamb) [inline packages/sdk/testcases/functional_tests/test_sdk_client.py]: Lets validate result here as well

File: packages/sdk/testcases/functional_tests/test_sdk_client.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    assert channels[0].state['last_signed_amount'] == 2000
+    assert channels[1].channel_id == 1
+    assert channels[1].state['nonce'] == 0
+    assert channels[1].state['last_signed_amount'] == 0
+
+
+def first_call_to_service_after_opening_second_channel(service_client):
+    request = examples_service_pb2.Numbers(a=20, b=3)
+
+    result = service_client.service.mul(request)
```
Comment (raamb) [inline packages/sdk/testcases/functional_tests/test_sdk_client.py]: Lets validate reult

File: packages/sdk/testcases/functional_tests/test_sdk_client.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    verify_when_no_open_channel(service_client)
+    open_first_channel(service_client)
+    first_call_to_service_after_opening_first_channel(service_client)
+    verify_channel_state_after_opening_first_channel_and_first_call_to_service(service_client)
+    second_call_to_service_after_opening_first_channel(service_client)
+    verify_channel_state_after_opening_first_channel_and_second_call_to_service(service_client)
+    open_second_channel(service_client)
+    verify_number_of_channel_after_opening_second_channel(service_client)
+    first_call_to_service_after_opening_second_channel(service_client)
+    verify_channel_state_after_opening_second_channel_and_first_call_to_service(service_client)
```
Comment (raamb) [inline packages/sdk/testcases/functional_tests/test_sdk_client.py]: Can we consider the scenario of making a call after a claim? This is to test service invocation when the nonce has changed

File: packages/snet_cli/snet/snet_cli/utils.py [hunk 10]
```diff
@@ -259,20 +259,28 @@ def rgetattr(obj, attr):
...
-        txn_hash = networks[w3.version.network]["transactionHash"]
-    return w3.eth.getTransactionReceipt(txn_hash).blockNumber
+    try:
+        with open(RESOURCES_PATH.joinpath("contracts", "networks", contract_file)) as f:
+            networks = json.load(f)
+            txn_hash = networks[w3.version.network]["transactionHash"]
+        return w3.eth.getTransactionReceipt(txn_hash).blockNumber
+    except Exception:
+        # TODO Hack as currenlty dependecy is on snet-cli so for test purpose return 0,need to remove dependecies from snet-cli ,currently very tightly coupled with it
+        if w3.version.network == 1 or w3.version.network == 3:
```
Comment (raamb) [inline packages/snet_cli/snet/snet_cli/utils.py]: Lets include Kovan network ID 42 as well till we resolve the hack

File: packages/sdk/snet/sdk/__init__.py (hunk 1 above)
Comment (keshrisohit) [reply packages/sdk/snet/sdk/__init__.py]: no its not will revisit this

File: packages/sdk/snet/sdk/service_client.py (hunk 5 above)
Comment (raamb) [reply packages/sdk/snet/sdk/service_client.py]: Looks like the behaviour is the same in snet-cli as well. Resolve for now but we need to track this.

File: packages/sdk/testcases/functional_tests/test_sdk_client.py (hunk 9 above)
Comment (raamb) [reply packages/sdk/testcases/functional_tests/test_sdk_client.py]: Created #378

**Pull request merged**